rules/onchain_budget.go

```diff
@@ -0,0 +1,694 @@
+		// find an entry in the store corresponding to the given request
+		// ID, then we assume it has already been accounted for.
+		if len(reqBytes) == 0 {
+			return nil
+		}
+
+		// If we did find an entry, we can now delete it.
+		err = store.Del(ctx, reqPref)
+		if err != nil {
+			return err
+		}
+
+		var request onChainAction
+		if err := json.Unmarshal(reqBytes, &request); err != nil {
+			return err
+		}
+
+		// Now, we subtract the reqAmt from total pending amt and store
+		// it.
+		if pending.Amount < request.Amount {
+			return fmt.Errorf("total pending cannot be less than " +
+				"requested amount")
+		}
+		pending.Amount -= request.Amount
+
+		b, err := json.Marshal(pending)
+		if err != nil {
+			return err
+		}
+
+		err = store.Set(ctx, onChainPending, b)
+		if err != nil {
+			return err
+		}
+
+		// Finally, we add the amount to the total spent and persist it
+		// to the permanent store.
+		complete.Amount += request.Amount
+		b, err = json.Marshal(complete)
+		if err != nil {
+			return err
+		}
+
+		return store.Set(ctx, onChainSpent, b)
+	})
+}
+
+// getBudgetState fetches the current state of the budget by getting the total
+// amount along with the total pending amount.
+func (o *OnChainBudgetEnforcer) getBudgetState(ctx context.Context,
+	tx firewalldb.KVStoreTx) (*onChainState, *onChainState, error) {
+
+	var (
+		spentAmt   onChainState
+		pendingAmt onChainState
+	)
+
+	store := tx.Local()
+
+	spentBytes, err := store.Get(ctx, onChainSpent)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pendingBytes, err := store.Get(ctx, onChainPending)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(spentBytes) != 0 {
+		err := json.Unmarshal(spentBytes, &spentAmt)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if len(pendingBytes) != 0 {
+		err := json.Unmarshal(pendingBytes, &pendingAmt)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return &spentAmt, &pendingAmt, nil
+}
+
+// addPendingAmt updates the store with the new amount. The total pending amount
+// is updated, and a request specific entry is added so that we know how much to
+// adjust our budget by if this request where to succeed or fail.
+func (o *OnChainBudgetEnforcer) addPendingAmt(ctx context.Context,
+	tx firewalldb.KVStoreTx, pending *onChainState,
+	request *onChainAction, reqID string) error {
+
+	store := tx.Local()
+
+	b, err := json.Marshal(
+		&onChainState{Amount: pending.Amount + request.Amount},
+	)
+	if err != nil {
+		return err
+	}
+
+	err = store.Set(ctx, onChainPending, b)
+	if err != nil {
+		return err
+	}
+
+	b, err = json.Marshal(request)
+	if err != nil {
+		return err
+	}
+
+	// Construct the unique key for this request.
+	reqPref := fmt.Sprintf("%s:%s", onChainPending, reqID)
+
+	// Check that we don't already have a pending payment for this request.
+	reqBytes, err := store.Get(ctx, reqPref)
+	if err != nil {
+		return err
+	}
+
+	if len(reqBytes) != 0 {
```

Comment:
I think it makes sense to add a test case for this edge case in the `onchain_budget_test.go`, and test that this get´s triggered if we send another request with the same `ReqID` before processing the last one.

I think it also makes sense to add coverage that you can make 2 `OpenChannelSync` requests at the same time before either of them has received a response, as long as the `ReqID` is different.